src/main/java/io/cdap/plugin/jms/common/JMSConfig.java

```diff
@@ -0,0 +1,250 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.jms.common;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Macro;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.plugin.common.ReferencePluginConfig;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Holds configuration required for configuring {@link io.cdap.plugin.jms.source.JMSSourceUtils;} and
+ * {@link io.cdap.plugin.jms.sink.JMSBatchSink}.
+ */
+public class JMSConfig extends ReferencePluginConfig implements Serializable {
+  // Schema
+  public static final String MESSAGE_ID = "messageId";
+  public static final String MESSAGE_TIMESTAMP = "messageTimestamp";
+  public static final String CORRELATION_ID = "correlationId";
+  public static final String REPLY_TO = "replyTo";
+  public static final String DESTINATION = "destination";
+  public static final String DELIVERY_MODE = "deliveryNode";
+  public static final String REDELIVERED = "redelivered";
+  public static final String TYPE = "type";
+  public static final String EXPIRATION = "expiration";
+  public static final String PRIORITY = "priority";
+
+  // Params
+  public static final String NAME_CONNECTION_FACTORY = "connectionFactory";
+  public static final String NAME_JMS_USERNAME = "jmsUsername";
+  public static final String NAME_JMS_PASSWORD = "jmsPassword";
+  public static final String NAME_PROVIDER_URL = "providerUrl";
+  public static final String NAME_TYPE = "type";
+  public static final String NAME_DESTINATION = "destination";
+  public static final String NAME_JNDI_CONTEXT_FACTORY = "jndiContextFactory";
+  public static final String NAME_JNDI_USERNAME = "jndiUsername";
+  public static final String NAME_JNDI_PASSWORD = "jndiPassword";
+  public static final String NAME_MESSAGE_TYPE = "messageType";
+  public static final String NAME_SCHEMA = "schema";
+
+  @Name(NAME_CONNECTION_FACTORY)
+  @Description("Name of the connection factory.")
+  @Nullable
+  @Macro
+  private String connectionFactory;
+
+  @Name(NAME_JMS_USERNAME)
+  @Description("Username to connect to JMS.")
+  @Macro
+  private String jmsUsername;
+
+  @Name(NAME_JMS_PASSWORD)
+  @Description("Password to connect to JMS.")
+  @Macro
+  private String jmsPassword;
+
+  @Name(NAME_PROVIDER_URL)
+  @Description("The URL of the JMS provider. For example, in case of an ActiveMQ Provider, the URL has the format " +
+    "tcp://hostname:61616.")
+  @Macro
+  private String providerUrl;
+
+  @Name(NAME_TYPE)
+  @Description("Queue or Topic.")
+  private String type;
```

Comment:
Does this need to be required? Since we already have a default for it?

Reply:
Can this property not have macro?

Reply:
No this doesn't need to be required. Added `@Nullable`.
Yes this property should have the `@Macro` annotation specified.